File: app/src/main/java/dev/bee/kanjianki/data/LocalStoreStudyStatus.java
package dev.bee.kanjianki.data;

import android.database.Cursor;

import java.util.ArrayList;
import java.util.List;

final class LocalStoreStudyStatus {
    private final LocalStore store;

    LocalStoreStudyStatus(LocalStore store) {
        this.store = store;
    }

    List<String> consumedTokens() {
        List<String> tokens = new ArrayList<>();
        try (Cursor cursor = store.getReadableDatabase().query(LocalStoreBase.TABLE_REVIEW_LOG, new String[]{LocalStoreBase.COLUMN_TOKEN}, null, null, null, null, null)) {
            while (cursor.moveToNext()) {
                tokens.add(LocalStoreBase.string(cursor, LocalStoreBase.COLUMN_TOKEN));
            }
        }
        return tokens;
    }

    LocalStoreBase.SyncStatus latestSync() {
        try (Cursor cursor = store.getReadableDatabase().query(LocalStoreBase.TABLE_SYNC_RUNS, null, null, null, null, null, LocalStoreBase.ORDER_ID_DESC, "1")) {
            if (!cursor.moveToFirst()) {
                return null;
            }
            return new LocalStoreBase.SyncStatus(new LocalStoreBase.SyncStatusValues(
                    LocalStoreBase.string(cursor, LocalStoreBase.COLUMN_STATUS),
                    LocalStoreBase.integer(cursor, LocalStoreBase.COLUMN_ACTIVE_NOTES_COUNT),
                    LocalStoreBase.integer(cursor, LocalStoreBase.COLUMN_ACTIVE_CARDS_COUNT),
                    LocalStoreBase.integer(cursor, LocalStoreBase.COLUMN_SUSPENDED_CARDS_ARCHIVED_COUNT),
                    LocalStoreBase.integer(cursor, LocalStoreBase.COLUMN_SUSPENDED_KANJI_IMPORTED_COUNT),
                    LocalStoreBase.longValue(cursor, LocalStoreBase.COLUMN_FINISHED_AT),
                    LocalStoreBase.string(cursor, LocalStoreBase.COLUMN_ERROR_MESSAGE),
                    LocalStoreBase.string(cursor, LocalStoreBase.COLUMN_REMOVAL_MESSAGE)
            ));
        }
    }

    boolean hasSuccessfulSyncSince(long finishedAtMillis) {
        try (Cursor cursor = store.getReadableDatabase().query(
                LocalStoreBase.TABLE_SYNC_RUNS,
                new String[]{"id"},
                "status=? AND finished_at>=?",
                new String[]{LocalStoreBase.STATUS_SUCCESS, Long.toString(finishedAtMillis)},
                null,
                null,
                LocalStoreBase.ORDER_ID_DESC,
                "1"
        )) {
            return cursor.moveToFirst();
        }
    }
}

File: app/src/main/kotlin/dev/bee/kanjianki/data/LocalStoreStudyStatus.kt
package dev.bee.kanjianki.data

internal class LocalStoreStudyStatus(
    private val store: LocalStore,
) {
    /**
     * Indexed single-token idempotency check. The review submit path used to load the
     * entire (unbounded, ever-growing) review_log token column on every Pass/Fail tap
     * just to test membership of one token; the token column is UNIQUE, so a targeted
     * lookup answers the same question in O(log n).
     */
    fun hasConsumedToken(token: String): Boolean {
        store.readableDatabase.query(
            LocalStoreBase.TABLE_REVIEW_LOG,
            arrayOf(LocalStoreBase.COLUMN_TOKEN),
            "${LocalStoreBase.COLUMN_TOKEN}=?",
            arrayOf(token),
            null,
            null,
            null,
            "1",
        ).use { cursor ->
            return cursor.moveToFirst()
        }
    }

    fun consumedTokens(): List<String> {
        val tokens = mutableListOf<String>()
        store.readableDatabase.query(
            LocalStoreBase.TABLE_REVIEW_LOG,
            arrayOf(LocalStoreBase.COLUMN_TOKEN),
            null,
            null,
            null,
            null,
            null,
        ).use { cursor ->
            while (cursor.moveToNext()) {
                tokens.add(LocalStoreBase.string(cursor, LocalStoreBase.COLUMN_TOKEN))
            }
        }
        return tokens
    }

    fun latestSync(): LocalStoreBase.SyncStatus? {
        store.readableDatabase.query(
            LocalStoreBase.TABLE_SYNC_RUNS,
            null,
            null,
            null,
            null,
            null,
            LocalStoreBase.ORDER_ID_DESC,
            "1",
        ).use { cursor ->
            if (!cursor.moveToFirst()) {
                return null
            }
            return LocalStoreBase.SyncStatus(
                LocalStoreBase.SyncStatusValues(
                    LocalStoreBase.string(cursor, LocalStoreBase.COLUMN_STATUS),
                    LocalStoreBase.integer(cursor, LocalStoreBase.COLUMN_ACTIVE_NOTES_COUNT),
                    LocalStoreBase.integer(cursor, LocalStoreBase.COLUMN_ACTIVE_CARDS_COUNT),
                    LocalStoreBase.integer(cursor, LocalStoreBase.COLUMN_SUSPENDED_CARDS_ARCHIVED_COUNT),
                    LocalStoreBase.integer(cursor, LocalStoreBase.COLUMN_SUSPENDED_KANJI_IMPORTED_COUNT),
                    LocalStoreBase.longValue(cursor, LocalStoreBase.COLUMN_FINISHED_AT),
                    LocalStoreBase.string(cursor, LocalStoreBase.COLUMN_ERROR_MESSAGE),
                    LocalStoreBase.string(cursor, LocalStoreBase.COLUMN_REMOVAL_MESSAGE),
                )
            )
        }
    }

    fun hasSuccessfulSyncSince(finishedAtMillis: Long): Boolean {
        store.readableDatabase.query(
            LocalStoreBase.TABLE_SYNC_RUNS,
            arrayOf("id"),
            "status=? AND finished_at>=?",
            arrayOf(LocalStoreBase.STATUS_SUCCESS, finishedAtMillis.toString()),
            null,
            null,
            LocalStoreBase.ORDER_ID_DESC,
            "1",
        ).use { cursor ->
            return cursor.moveToFirst()
        }
    }
}
